datastack: Improve show_stack and add _repr_html_ for stack

hamogu opened the pull request:
# Summary
Improve datastack display (via `show_stack` and in the notebook) for datastacks.

# Details
The previous code listed only the OBSID and MJD_OBS and the name of the datstack. This works well when using a datstack for a time series of spectra. However, for other use cases where all spectra in the datstack from from the observation (and sometimes even form the same PHA II file) that display is useless. In particular for grating data (where HEG and MEG positive and negative orders all have the same OBSID and MJD_OBS)  or polarimetry data (IXPE) all displayed values are identical - not very useful.
Now, the table is constructed from a longer list of header keywords, but only keywords that are present and actually differ between rows are displayed; the list of keywords used can be changed as a module-level variable if needed.
I also added a html represenation for notebook display.

# Notes
The previous implementation went to great length to unify treatment for files that have MJD_OBS or MJD-OBS set. CIAO changed that header keywords a few years ago. I don't think that complexity is warranted any longer. I just treat the columns separately. In practice, cases where people mix files from a 3 year old CIAO version with a current CIAO version in the same datastack are going to be rare and columns with non-existent keywords are not displayed, so users will see either MJD-OBD or MJD_OBS but not both.

# Example
old:
![Screen Shot 2021-06-20 at 6 51 43 AM](https://user-images.githubusercontent.com/498688/122671646-b318a080-d195-11eb-95e5-c24ad23b1ec9.png)

new:
![Screen Shot 2021-06-20 at 6 52 04 AM](https://user-images.githubusercontent.com/498688/122671652-b9a71800-d195-11eb-9248-ece97d3fe36e.png)

wmclaugh commented on sherpa/astro/datastack/ds.py:
```diff
@@ -108,47 +121,70 @@ def clear_stack(self):
+        header : list of strings
+            Column names
+        data : list of lists
+            Eahc list is a row, corresponding to one dataset
```
Eahc -> Each

wmclaugh commented on sherpa/astro/datastack/tests/test_datastack.py:
```diff
@@ -723,134 +724,60 @@ def test_default_instantiation(ds_setup):
+    assert len(lines) == 2
+    assert 'empty datastack' in lines[0]
+    assert lines[1].strip() == ''
+     # Now, the check the html represention
```
represention -> representation

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -366,7 +366,8 @@ def html_pha(pha):
                           ('HDUCLAS3', 'Data format'),
-                          ('HDUCLAS4', 'PHA format')])
+                          ('HDUCLAS4', 'PHA format'),
+                          ('XFLT0001', 'Stokes (polarimetry)')])
```
Technically I think that the `XFLTxxxx` keywords are generic - that is you could have one in a non-polarimetry dataset.

DougBurke replied:
So I'm not sure what we do here.

DougBurke commented on sherpa/astro/datastack/ds.py:
```diff
@@ -108,47 +121,70 @@ def clear_stack(self):
                 hdr = dataset['data'].header
+                row.extend([hdr.get(k, '') for k in keys_to_display])
+            except AttributeError:
+                row.append([''] * len(keys_to_display))
```
I'm guessing this is to check for obects which don't have a header? Your recent changes mean that we don't getthis by default, but I guess a user could do something like `pha.header = None`.

DougBurke replied:
[This is not a request to change code, just so I can understand it as there's no test coverage]

hamogu replied:
Yes.

DougBurke commented on sherpa/astro/datastack/ds.py:
```diff
@@ -108,47 +121,70 @@ def clear_stack(self):
+        header, data = self._summary_for_display()
+        print('|'.join(header))
+        for row in data:
+            print('|'.join(row))
```
This has a different feel to the other `show_xxx` routines, which tend to use more spaces, but I think the existing routines have a varied enough approach that we don't have a consistent guide.

DougBurke replied:
[a general comment, not a request for changes]

DougBurke commented on sherpa/astro/datastack/ds.py:
```diff
@@ -108,47 +121,70 @@ def clear_stack(self):
 
-            except AttributeError:
-                pass
+    # QUS: should this use the logger instead of print?
```
The existing `show_xxx` routines all use `sherpa.utils.send_to_pager` [at least those in sherpa[.astro].ui.utils do] so we should probly do the same.  I'm not a great fan of `send_to_pager` but its what we do...

DougBurke replied:
Of course, I don't know how the `show_xxx` routines work in the noteook/lab interface.

hamogu replied:
This is not new code, it's merely moved to a different line: https://github.com/sherpa/sherpa/pull/1194/files#diff-1b2333a505749af538cd4ba44a468fce606ee4cea032f1e6544ce66aa85b3debL111

This `shows_stack` function did not `send_to_pager` before and it does not do so now. All I have changed in the content of the message. (And there is some refactoring, because `_repr_html` needs similar information.
So, just to clarify: Do you think I should bring is in-line with what the other `show_xxx` functions do because I'm touching this function now?
I don't know what the original motivations was to use or not use `send_to_pager` in `show_stack`.

DougBurke reviewed (changes requested):
Deciding what to do with XFLT and whether send_to_pager works in the notebooks are the issues I see.

hamogu commented:
I've corrected the label for XFLT and modified `show_stack` to use `send_to_pager` in line with what the other `show_xxx` functions do, addressing all points from the review.

hamogu commented:
The red checkbox is only a marginal decrease in codecov. I believe that this PR is ready for re-review by @DougBurke and (if approved) can be merged regardless.

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -366,7 +366,8 @@ def html_pha(pha):
                           ('HDUCLAS3', 'Data format'),
-                          ('HDUCLAS4', 'PHA format')])
+                          ('HDUCLAS4', 'PHA format'),
+                          ('XFLT0001', 'XEPC filter 0001')])
```
```suggestion
                          ('XFLT0001', 'XSPEC filter 0001')])
```

DougBurke commented on sherpa/astro/datastack/tests/test_datastack.py:
```diff
@@ -723,134 +724,60 @@ def test_default_instantiation(ds_setup):
+def test_show_empty_datastack(capsys):
+    '''Test output for an empty datastack.
 
+    For a change, we try the OO interface here/
```
```suggestion
    For a change, we try the OO interface here.
```

DougBurke replied:
The mostest important fix.

DougBurke reviewed (changes requested):
Just the two typos.

The original datastack code was developed outside Sherpa by @taldcroft and then brought into Sherpa so I imagine we never really looked at how `show_stack` was implemented when we did so.

hamogu commented:
Fixed the typos according to @DougBurke suggestions about 30 s ago, so this is hopefully ready. Asking @DougBurke for final review.

DougBurke commented:
Just waiting for the tests to pass.

hamogu commented:
Rebased to pick up recent changes in main.

DougBurke commented:
Why are these tests still waiting to run? eg https://github.com/sherpa/sherpa/actions/runs/1072355480

DougBurke commented:
```
Can't find any online and idle self-hosted or hosted runner in the current repository, account/organization that matches the required labels: 'ubuntu-latest'
Waiting for a self-hosted or a hosted runner to pickup this job...
```

That doesn't look great...I'm going to cancel the conda runs (as none of them have run) and see if we can restart them, assuming this is a transient failure on GitHub's side.
